napalm_srlinux/srlinux.py

```diff
@@ -0,0 +1,1036 @@
+        )
+
+        if ok:
+            return response.get("result")
+
+        raise Exception(response.get("error", {}).get("message"))
+
+    def _jsonrpc_request(self, method: RPCMethod, params: dict) -> tuple[bool, dict]:
+        """
+        Make a JSON RPC request, raise an exception if the HTTP request returns anything other than 2xx
+        Return a boolean success value (if HTTP 2xx) and the result.
+        """
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+
+        request_data = {
+            "jsonrpc": "2.0",
+            "id": datetime.datetime.now().strftime("%s"),
+            "method": method,
+            "params": params,
+        }
+
+        result = self.jsonrpc_client.post(
+            self.jsonrpc_url, headers=headers, json=request_data, timeout=self.timeout
+        )
+
+        if result.status_code == httpx.codes.OK and result.json().get("error"):
+            return False, result.json()
+        elif result.status_code == httpx.codes.OK:
+            return True, result.json()
+        elif result.status_code == httpx.codes.BAD_REQUEST:
+            raise Exception("Request raised HTTP/400 BAD REQUEST")
+        else:
+            raise Exception(f"Request raised unknown status code {result.status_code}")
+
+    def _new_jsonrpc_client(self):
+        """
+        Create a JSON RPC Client, preconfigured with TLS if required.
+        """
+        cert = None
+        if not self.insecure:
+            cert = (
+                (self.tls_cert_path, self.tls_key_path, self.tls_key_password)
+                if self.tls_key_password
+                else (self.tls_cert_path, self.tls_key_path)
+            )
+
+        opts = {"verify": (not self.insecure), "auth": (self.username, self.password)}
+
+        if cert:
+            opts["cert"] = cert
+
+        return httpx.Client(**opts)
+
+    def _compose_jsonrpc_url(self):
+        """
+        Compose the JSON RPC URL, based on the initialized arguments.
+        """
+
+        proto = (
+            "https"
+            if (
+                self.jsonrpc_port == 443
+                or (self.jsonrpc_port != 80 and not self.insecure)
+            )
+            else "http"
+        )
+        url = f"{proto}://{self.hostname}:{self.jsonrpc_port}/jsonrpc"
+        # print(url)
+
+        return url
+
+    def _determine_jsonrpc_port(self, opt_args):
+        """
+        Determine the JSON RPC port based on the optional arguments.
+        """
+
+        # by default assume 443 port is used by jsonrpc server
+        port: int = 443
+
+        # if jsonrpc_port is specified, use that
+        if opt_args.get("jsonrpc_port"):
+            port = opt_args.get("jsonrpc_port")
+        # when insecure is set and jsonrpc_port is not specified, use 80
+        elif opt_args.get("insecure"):
+            port = 80
+
+        return port
```

Comment:
I added these two helper functions to have some flexibility and testability on how we determine the jsonrpc port and the resulting URL

Reply:
Excellent. There’s a few other methods throughout the place that I think could do with the same treatment. These methods are also a good place to start for unit tests.

Reply:
also added a dumb unit test (which fails for now since the driver wants to have some cert files which we don't have yet in the mock data)